dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/MemToReg.java

```diff
@@ -0,0 +1,353 @@
+        for (final Alloc allocation : function.getEvents(Alloc.class)) {
+            matcher.reachabilityGraph.put(allocation, new HashSet<>());
+        }
+        // This loop should terminate, since back jumps occur, only if changes were made.
+        Label back;
+        for (int i = 0; i < events.size(); i = back == null ? i + 1 : Collections.binarySearch(events, back, eventComparator)) {
+            back = events.get(i).accept(matcher);
+        }
+        // Replace every unmarked address.
+        final var replacingRegisters = new HashMap<RegWriter, List<Register>>();
+        for (final Alloc allocation : function.getEvents(Alloc.class)) {
+            if (matcher.reachabilityGraph.containsKey(allocation)) {
+                final List<Type> registers = getReplacingRegisters(allocation);
+                if (registers != null) {
+                    replacingRegisters.put(allocation, registers.stream().map(function::newRegister).toList());
+                    assert allocation.getUsers().isEmpty();
+                    allocation.tryDelete();
+                }
+            }
+        }
+        int loadCount = 0, storeCount = 0;
+        // Replace all loads and stores to replaceable storage.
+        for (final Map.Entry<MemoryEvent, AddressOffset> entry : matcher.accesses.entrySet()) {
+            final MemoryEvent event = entry.getKey();
+            final AddressOffset access = entry.getValue();
+            final List<Register> registers = access == null ? null : replacingRegisters.get(access.base);
+            if (registers == null || access.offset < 0 || access.offset >= registers.size()) {
+                continue;
+            }
+            if (event instanceof Load load) {
+                final Register reg = load.getResultRegister();
+                assert load.getUsers().isEmpty();
+                load.replaceBy(EventFactory.newLocal(reg, expressions.makeCast(registers.get(access.offset), reg.getType())));
+                loadCount++;
+            }
+            if (event instanceof Store store) {
+                final Register reg = registers.get(access.offset);
+                assert store.getUsers().isEmpty();
+                store.replaceBy(EventFactory.newLocal(reg, expressions.makeCast(store.getMemValue(), reg.getType())));
+                storeCount++;
+            }
+        }
+        // Remove involved local assignments.
+        for (final Map.Entry<Local, AddressOffset> entry : matcher.assignments.entrySet()) {
+            if (replacingRegisters.containsKey(entry.getValue().base)) {
+                assert entry.getKey().getUsers().isEmpty();
+                entry.getKey().tryDelete();
+            }
+        }
+        if (loadCount + storeCount > 0) {
+            logger.debug("Removed {} loads and {} stores in function \"{}\".", loadCount, storeCount, function.getName());
+        }
+    }
+
+    private List<Type> getReplacingRegisters(Alloc allocation) {
```

Comment:
I think the name is not fitting because this method does not return registers. 
`getReplacementTypes` might be better.
If the method is intended to only return primitive/simple types (non-aggregates), you can also call it `getPrimitiveReplacementTypes`.